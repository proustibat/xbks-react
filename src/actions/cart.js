export const addBook = book => ( {
    type: 'ADD_BOOK',
    book
} );

export const removeBook = ( isbn = '' ) => ( {
    type: 'REMOVE_BOOK',
    isbn
} );

// Could be used if cart is saved in localstorage or cookie
export const setSavedBooks = books => ( {
    type: 'SET_SAVE_BOOKS',
    books
} );

export const changeQuantity = ( { isbn, qtt } ) => ( {
    type: 'CHANGE_QUANTITY',
    isbn,
    qtt
} );
