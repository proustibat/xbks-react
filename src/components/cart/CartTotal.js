import React from 'react';
import { connect } from 'react-redux';
import getSubTotal from '../../selectors/getSubTotal';
import getBestDiscount from '../../selectors/getBestDiscount';
import numeral from "numeral";

const CartTotal = props => (
    <div className="cart-totals">
        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Total before reductions:</span>
            <span className="cart-totals__line-amount">{ numeral( props.subtotal ).format( '0.00$' ) }</span>
        </p>
        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Reductions <em>(type: { props.discount.type })</em>:</span>
            <span className="cart-totals__line-amount">{ numeral( props.discount.amount ).format( '0.00$' ) }</span>
        </p>

        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Total to pay:</span>
            <span className="cart-totals__line-amount">{ numeral( props.total ).format( '0.00$' ) }</span>
        </p>
    </div>
);

const mapStateToProps = state => ( {
    subtotal: getSubTotal( state.cart.books ),
    discount: state.offers.bestOffer,
    total: getSubTotal( state.cart.books ) - state.offers.bestOffer.amount
} );

export default connect( mapStateToProps )( CartTotal );
