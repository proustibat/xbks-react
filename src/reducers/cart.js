const cartReducerDefaultState = {
    books: []
};

export default ( state = cartReducerDefaultState, action ) => {
    let books;
    switch ( action.type ) {
        case 'ADD_BOOK_TO_CART':
            const existingBook = state.books.find( book => book.isbn === action.book.isbn );
            if ( existingBook ) {
                existingBook.quantity = existingBook.quantity + 1;
                books = state.books;
            }
            else {
                action.book.quantity = 1;
                books = [ ...state.books, action.book ];
            }
            return { ...state, ...{ books } };
        case 'REMOVE_BOOK_FROM_CART':
            books = state.books.filter( ( { isbn } ) => action.isbn !== isbn );
            return { ...state, ...{ books } };
        // case 'SET_SAVED_BOOKS_TO_CART':
        //     action.books.forEach( book => {
        //         if ( !book.quantity ) {
        //             book.quantity = 1;
        //         }
        //     } );
        //     return { ...state, ...{ books: action.books } };
        case 'CHANGE_QUANTITY_BOOK_IN_CART':
            const booksList = [...state.books];
            let newBookList = booksList
                .map( book => {
                    book.quantity = book.isbn === action.isbn ? action.qtt : book.quantity;
                    return book;
                } )
                .filter( book => book.quantity > 0 );
            return { ...state, ...{ books: newBookList } };
        default:
            return state;
    }
};
