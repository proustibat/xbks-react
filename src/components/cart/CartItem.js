import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import { changeQuantity, removeBook } from '../../actions/cart';
import { startSetOffers } from "../../actions/offers";

export class CartItem extends React.Component {

    state = {
      subTotal: 0
    };

    constructor( props ) {
        super( props );
        this.state = {
            isbn: props.isbn || '',
            title: props.title || '',
            price: props.price || 0,
            quantity: props.quantity || 0
        }
    }

    handleNumberChange = ( e ) => {
        console.log( this.props.isbn, e.currentTarget.value );
        const quantity = parseInt( e.currentTarget.value, 10 );

        this.setState( () => ( { quantity } ) );

        this.props.changeQuantity( {
            isbn: this.state.isbn,
            qtt: quantity
        } );
        this.props.startSetOffers();
    };

    handleRemove = () => {
      this.props.removeBook( this.state.isbn );
      this.props.startSetOffers();
    };

    render() {
        return (
            <div className="cart-list-item">
            <span className="cart-list__col-item">
                <Link className="cart-list__col-item-title" to={ `/book/${ this.state.isbn }` }>{ this.state.title }</Link>
            </span>
            <span className="cart-list__col-qtt">
                <button className="cart-list__delete" onClick={ this.handleRemove }>Remove</button>
                <input type="number" min="0" id="itemsNumber" defaultValue={ this.state.quantity } step="1" onInput={ this.handleNumberChange } />
            </span>
                <span className="cart-list__col-unit-price">{ numeral( this.state.price ).format( '0.00$') }</span>
                <span className="cart-list__col-sub-total">{ numeral( this.state.quantity * this.state.price ).format( '0.00$' ) }</span>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ( {
    changeQuantity: ( { isbn, qtt } ) => dispatch( changeQuantity( { isbn, qtt } ) ),
    removeBook: isbn => dispatch( removeBook( isbn ) ),
    startSetOffers: () => dispatch ( startSetOffers() )
} );

export default connect( undefined, mapDispatchToProps )( CartItem );
