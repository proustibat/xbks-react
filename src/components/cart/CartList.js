import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';

const CartList = props => (
    <div className="cart-list">
        { props.books.length === 0 ? (
            <span>Your cart is empty</span>
        ) : (
            <div>
                <div className="cart-list-header">
                    <span className="cart-list__col-item">Item</span>
                    <span className="cart-list__col-qtt">Qtt</span>
                    <span className="cart-list__col-unit-price">Unit Price</span>
                    <span className="cart-list__col-sub-total">Sub Total</span>
                </div>
                <div className="cart-list-body">
                    { props.books.map( ( book, i ) => (
                        <CartItem key = { `${ book.isbn }-${ i }` } { ...book } />
                    ) ) }
                </div>
            </div>
        ) }
    </div>
);

const mapStateToProps = state => ( {
    books: state.cart.books
} );

export default connect( mapStateToProps )( CartList);
