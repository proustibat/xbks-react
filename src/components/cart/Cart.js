import React from 'react';
import CartList from './CartList';
import CartTotal from './CartTotal';

export class Cart extends React.Component {
    render () {
        return (
            <div className="cart">
                <CartList />
                <CartTotal />
            </div>
        );
    }
}

export default Cart;
