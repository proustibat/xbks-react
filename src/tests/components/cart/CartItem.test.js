import React from 'react';
import { shallow } from 'enzyme';
import { CartItem } from "../../../components/cart/CartItem";
import cart from "../../fixtures/cart";

test( 'Should render CartItem component correctly with data', () => {
    const itemData = cart.books[ 0 ];
    const wrapper = shallow( <CartItem
        { ...itemData }
    /> );
    expect( wrapper ).toMatchSnapshot();
} );


test( 'Should render CartItem component correctly with no data', () => {
    const wrapper = shallow( <CartItem /> );
    expect( wrapper ).toMatchSnapshot();
} );

test( 'Should handle change number of quantity input (update state, update offers and quantity in store)', () => {
    const itemData = cart.books[ 0 ];
    const startSetOffers = jest.fn();
    const changeQuantity = jest.fn();
    const wrapper = shallow( <CartItem
        { ...itemData }
        startSetOffers = { startSetOffers }
        changeQuantity = { changeQuantity }
    /> );
    const value = 333;
    wrapper.find( '.cart-list__col-qtt input' ).first().simulate( 'input', { currentTarget: { value } } );

    expect( wrapper.state( 'quantity' ) ).toBe( value );
    expect( changeQuantity ).toHaveBeenLastCalledWith( {
        isbn: wrapper.state( 'isbn' ),
        qtt: value
    } );
    expect( startSetOffers ).toHaveBeenCalled();
} );

test( 'Should handle remove book', () => {
    const itemData = cart.books[ 0 ];
    const startSetOffers = jest.fn();
    const removeBook = jest.fn();
    const wrapper = shallow( <CartItem
        { ...itemData }
        startSetOffers = { startSetOffers }
        removeBook = { removeBook }
    /> );

    wrapper.find( '.cart-list__col-qtt button' ).first().simulate( 'click' );

    expect( removeBook ).toHaveBeenLastCalledWith( wrapper.state( 'isbn' ) );
    expect( startSetOffers ).toHaveBeenCalled();
} );
