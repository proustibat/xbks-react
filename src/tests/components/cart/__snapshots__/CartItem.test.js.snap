// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CartItem component correctly with data 1`] = `
<div
  className="cart-list-item"
>
  <span
    className="cart-list__col-item"
  >
    <Link
      className="cart-list__col-item-title"
      replace={false}
      to="/book/c8fabf68-8374-48fe-a7ea-a00ccd07afff"
    >
      Henri Potier à l'école des sorciers
    </Link>
  </span>
  <span
    className="cart-list__col-qtt"
  >
    <button
      className="cart-list__delete"
      onClick={[Function]}
    >
      Remove
    </button>
    <input
      defaultValue={1}
      id="itemsNumber"
      min="0"
      onInput={[Function]}
      step="1"
      type="number"
    />
  </span>
  <span
    className="cart-list__col-unit-price"
  >
    35.00$
  </span>
  <span
    className="cart-list__col-sub-total"
  >
    35.00$
  </span>
</div>
`;

exports[`Should render CartItem component correctly with no data 1`] = `
<div
  className="cart-list-item"
>
  <span
    className="cart-list__col-item"
  >
    <Link
      className="cart-list__col-item-title"
      replace={false}
      to="/book/"
    />
  </span>
  <span
    className="cart-list__col-qtt"
  >
    <button
      className="cart-list__delete"
      onClick={[Function]}
    >
      Remove
    </button>
    <input
      defaultValue={0}
      id="itemsNumber"
      min="0"
      onInput={[Function]}
      step="1"
      type="number"
    />
  </span>
  <span
    className="cart-list__col-unit-price"
  >
    0.00$
  </span>
  <span
    className="cart-list__col-sub-total"
  >
    0.00$
  </span>
</div>
`;
